firefox_puppeteer/tests/test_windows.py

```diff
@@ -3,98 +3,181 @@
-                          self.marionette.current_chrome_window_handle)
-
+        # Switch to the last window and find 2nd window by URL
         self.windows.switch_to(windows[2].handle)
-        self.assertEquals(windows[2].handle,
-                          self.marionette.current_chrome_window_handle)
 
-        def is_my_window(win):
-            with self.marionette.using_context('content'):
-                return self.marionette.get_url() == url
+        def find_by_url(win):
+            with win.marionette.using_context('content'):
+                return win.marionette.get_url() == url
 
-        self.windows.switch_to(is_my_window)
-        self.assertEquals(windows[0].handle,
-                          self.marionette.current_chrome_window_handle)
+        self.windows.switch_to(find_by_url)
+        self.assertEquals(windows[1].handle, self.marionette.current_chrome_window_handle)
 
-        with self.assertRaises(NoSuchElementException):
-            self.windows.switch_to("humbug")
+        # Switching to an unknown handles has to fail
+        self.assertRaises(errors.NoSuchWindowException,
+                          self.windows.switch_to, "humbug")
+        self.assertRaises(errors.NoSuchWindowException,
+                          self.windows.switch_to, lambda win: False)
 
-        with self.assertRaises(NoSuchElementException):
-            self.windows.switch_to(lambda win: False)
+        self.windows.close_all([self.browser])
+        self.browser.switch_to()
 
-        windows[1].close()
-        windows[2].close()
+        self.assertEqual(len(self.windows.all), 1)
 
-    def test_base_window(self):
-        self.assertEquals(len(self.windows.all), 1)
-        self.assertEquals(len(self.marionette.chrome_window_handles), 1)
+    def test_base_window_basics(self):
+        self.browser.__class__ = BaseWindow  # force BaseWindow instance
 
-        first_window = self.windows.current
-        self.assertEquals(first_window.handle,
+        self.assertEquals(self.browser.handle,
                           self.marionette.current_chrome_window_handle)
-        self.assertEquals(first_window.get_attribute('windowtype'),
+        self.assertEquals(self.browser.window,
+                          self.marionette.find_element(By.CSS_SELECTOR, ':root'))
+        self.assertEquals(self.browser.window.get_attribute('windowtype'),
                           self.marionette.get_window_type())
-        self.assertEquals(first_window.window,
-                          self.marionette.find_element(By.CSS_SELECTOR,
-                                                       ':root'))
+        self.assertFalse(self.browser.closed)
 
-        # This action will fail if the current window is not a browser window
-        first_window.send_shortcut('ctrl-n')
+    def test_base_window_open_close(self):
+        win1 = self.windows.current
+        win1.__class__ = BaseWindow  # force BaseWindow instance
```

Comment:
I don't know how this is meant to work, but I'd like to find another way to achieve this if we can.